Origin: https://github.com/php/php-src/commit/ed70b1ea43a9b7ffa2f53b3e5d6ba403f37ae81c
Advisory: GHSA-www2-q4fc-65wf
Imported-by: mammoth-lts 7.2 backport (patches/7.2/9003), adapted to 7.1 (zend_parse_parameters() format strings)

From 3fce1f6ef2fe1939c77ca595190c24fef3070f46 Mon Sep 17 00:00:00 2001
From: Niels Dossche <7771979+nielsdos@users.noreply.github.com>
Date: Sat, 6 Sep 2025 21:55:13 +0200
Subject: [PATCH] Fix GHSA-www2-q4fc-65wf

Backport of php-src ed70b1ea43a9b7ffa2f53b3e5d6ba403f37ae81c (PHP-8.1) to PHP 7.2:
- getservbyname() name is a char* in 7.2: Z_PARAM_PATH instead of Z_PARAM_PATH_STR
- dns_win32.c: "s" -> "p" in the existing 7.2 format strings
- test: 7.2 ZPP failures are a warning + NULL (false for inet_pton) instead of ValueError
- existing tests updated: wrong-type warnings now say "to be a valid path" instead of "to be string"

7.1: the functions parse their arguments with zend_parse_parameters(): "s" -> "p" instead of
Z_PARAM_STRING -> Z_PARAM_PATH; tests as in the 7.2 port.
---
diff --git a/ext/standard/basic_functions.c b/ext/standard/basic_functions.c
index dac101109a382d5e0274197c856e4df4ee04e9b4..7b614f597139a84a10fa7b3784ed3ba5b850d311 100644
--- a/ext/standard/basic_functions.c
+++ b/ext/standard/basic_functions.c
@@ -3944,7 +3944,7 @@ PHP_NAMED_FUNCTION(php_inet_pton)
 	size_t address_len;
 	char buffer[17];
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &address, &address_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p", &address, &address_len) == FAILURE) {
 		RETURN_FALSE;
 	}
 
@@ -3984,7 +3984,7 @@ PHP_FUNCTION(ip2long)
 	zend_ulong ip;
 #endif
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &addr, &addr_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p", &addr, &addr_len) == FAILURE) {
 		return;
 	}
 
@@ -5634,7 +5634,7 @@ PHP_FUNCTION(getservbyname)
 	size_t name_len, proto_len;
 	struct servent *serv;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "ss", &name, &name_len, &proto, &proto_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pp", &name, &name_len, &proto, &proto_len) == FAILURE) {
 		return;
 	}
 
@@ -5677,7 +5677,7 @@ PHP_FUNCTION(getservbyport)
 	zend_long port;
 	struct servent *serv;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "ls", &port, &proto, &proto_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "lp", &port, &proto, &proto_len) == FAILURE) {
 		return;
 	}
 
@@ -5701,7 +5701,7 @@ PHP_FUNCTION(getprotobyname)
 	size_t name_len;
 	struct protoent *ent;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &name, &name_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p", &name, &name_len) == FAILURE) {
 		return;
 	}
 
diff --git a/ext/standard/dns.c b/ext/standard/dns.c
index b5fbcb96f968c83ecf75c39900cf7d88e19fff3a..654809be185451d356c708190c7f880f42854979 100644
--- a/ext/standard/dns.c
+++ b/ext/standard/dns.c
@@ -372,7 +372,7 @@ PHP_FUNCTION(dns_check_record)
 	struct __res_state *handle = &state;
 #endif
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|s", &hostname, &hostname_len, &rectype, &rectype_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p|s", &hostname, &hostname_len, &rectype, &rectype_len) == FAILURE) {
 		return;
 	}
 
@@ -817,7 +817,7 @@ PHP_FUNCTION(dns_get_record)
 	int type, first_query = 1, store_results = 1;
 	zend_bool raw = 0;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|lz/!z/!b",
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p|lz/!z/!b",
 			&hostname, &hostname_len, &type_param, &authns, &addtl, &raw) == FAILURE) {
 		return;
 	}
@@ -1048,7 +1048,7 @@ PHP_FUNCTION(dns_get_mx)
 	struct __res_state *handle = &state;
 #endif
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sz/|z/", &hostname, &hostname_len, &mx_list, &weight_list) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pz/|z/", &hostname, &hostname_len, &mx_list, &weight_list) == FAILURE) {
 		return;
 	}
 
diff --git a/ext/standard/dns_win32.c b/ext/standard/dns_win32.c
index a35f1656765fc50974e431462be592200fa6c031..d4fa864df0640172fe00b161ea03b1124a1bb5d0 100644
--- a/ext/standard/dns_win32.c
+++ b/ext/standard/dns_win32.c
@@ -50,7 +50,7 @@ PHP_FUNCTION(dns_get_mx) /* {{{ */
 	DNS_STATUS      status;                 /* Return value of DnsQuery_A() function */
 	PDNS_RECORD     pResult, pRec;          /* Pointer to DNS_RECORD structure */
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sz/|z/", &hostname, &hostname_len, &mx_list, &weight_list) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pz/|z/", &hostname, &hostname_len, &mx_list, &weight_list) == FAILURE) {
 		return;
 	}
 
@@ -99,7 +99,7 @@ PHP_FUNCTION(dns_check_record)
 	DNS_STATUS      status;                 /* Return value of DnsQuery_A() function */
 	PDNS_RECORD     pResult;          /* Pointer to DNS_RECORD structure */
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|s", &hostname, &hostname_len, &rectype, &rectype_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p|s", &hostname, &hostname_len, &rectype, &rectype_len) == FAILURE) {
 		return;
 	}
 
@@ -352,7 +352,7 @@ PHP_FUNCTION(dns_get_record)
 	int type, type_to_fetch, first_query = 1, store_results = 1;
 	zend_bool raw = 0;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|lz/!z/!b",
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p|lz/!z/!b",
 			&hostname, &hostname_len, &type_param, &authns, &addtl, &raw) == FAILURE) {
 		return;
 	}
diff --git a/ext/standard/tests/general_functions/getservbyname_variation1.phpt b/ext/standard/tests/general_functions/getservbyname_variation1.phpt
index 7dd01aab3679985c06156d58cca18e1806a655bc..ac7d9fe041433ac1175443108500e3cafc93a73d 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation1.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation1.phpt
@@ -27,14 +27,14 @@ foreach ( $variation_array as $var ) {
 --EXPECTF--
 *** Test substituting argument 1 with array values ***
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
diff --git a/ext/standard/tests/general_functions/getservbyname_variation13.phpt b/ext/standard/tests/general_functions/getservbyname_variation13.phpt
index 8dad8cda55fff4186b3d4bd508e2bb9a77ea443b..b556d4646d4fff42eb2644b276426bb74b73c4e1 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation13.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation13.phpt
@@ -34,5 +34,5 @@ foreach ( $variation_array as $var ) {
 *** Test substituting argument 2 with object values ***
 bool(false)
 
-Warning: getservbyname() expects parameter 2 to be string, object given in %s.php on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, object given in %s.php on line %d
 NULL
diff --git a/ext/standard/tests/general_functions/getservbyname_variation6.phpt b/ext/standard/tests/general_functions/getservbyname_variation6.phpt
index 0dfafa6229518f735585ad443697d2231788c0ab..e184c1a1c4ac5f6fa8ab3740078f626b8ec43099 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation6.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation6.phpt
@@ -34,5 +34,5 @@ foreach ( $variation_array as $var ) {
 *** Test substituting argument 1 with object values ***
 bool(false)
 
-Warning: getservbyname() expects parameter 1 to be string, object given in %s.php on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, object given in %s.php on line %d
 NULL
diff --git a/ext/standard/tests/general_functions/getservbyname_variation8.phpt b/ext/standard/tests/general_functions/getservbyname_variation8.phpt
index 69d1d77b6a7765e9278954b19dd1570319622e47..78d7d0b763366435e2cb6933d4a4aacfeb63134c 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation8.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation8.phpt
@@ -27,14 +27,14 @@ foreach ( $variation_array as $var ) {
 --EXPECTF--
 *** Test substituting argument 2 with array values ***
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
diff --git a/ext/standard/tests/network/ghsa-www2-q4fc-65wf.phpt b/ext/standard/tests/network/ghsa-www2-q4fc-65wf.phpt
new file mode 100644
index 0000000000000000000000000000000000000000..0f7cbe1eb0b378c942f32533db1b76b41cb334ee
--- /dev/null
+++ b/ext/standard/tests/network/ghsa-www2-q4fc-65wf.phpt
@@ -0,0 +1,43 @@
+--TEST--
+GHSA-www2-q4fc-65wf
+--DESCRIPTION--
+This is a ZPP test but *keep* this as it is security-sensitive!
+--FILE--
+<?php
+var_dump(dns_check_record("\0"));
+var_dump(dns_get_mx("\0", $out));
+var_dump(dns_get_record("\0"));
+var_dump(getprotobyname("\0"));
+var_dump(getservbyname("\0", "tcp"));
+var_dump(getservbyname("x", "tcp\0"));
+var_dump(getservbyport(0, "tcp\0"));
+var_dump(inet_pton("\0"));
+var_dump(ip2long("\0"));
+?>
+--EXPECTF--
+Warning: dns_check_record() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: dns_get_mx() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: dns_get_record() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getprotobyname() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getservbyname() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getservbyname() expects parameter 2 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getservbyport() expects parameter 2 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: inet_pton() expects parameter 1 to be a valid path, string given in %s on line %d
+bool(false)
+
+Warning: ip2long() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
diff --git a/ext/standard/tests/network/ip.phpt b/ext/standard/tests/network/ip.phpt
index 03329802c1cb048cbae42c9b376954f4a23824f6..8e36199cabfd61a691111cd6da61e045543c165e 100644
--- a/ext/standard/tests/network/ip.phpt
+++ b/ext/standard/tests/network/ip.phpt
@@ -54,7 +54,7 @@ bool(false)
 bool(false)
 int(1869573999)
 
-Warning: ip2long() expects parameter 1 to be string, array given in %sip.php on line %d
+Warning: ip2long() expects parameter 1 to be a valid path, array given in %sip.php on line %d
 NULL
 
 Warning: long2ip() expects exactly 1 parameter, 0 given in %sip.php on line %d
diff --git a/ext/standard/tests/network/ip2long_variation1.phpt b/ext/standard/tests/network/ip2long_variation1.phpt
index fa7410930f314d59aa508d10492a89c2de3ebd35..b305ae4151fa9237a5ba1bbc65e4b76fe3cf303a 100644
--- a/ext/standard/tests/network/ip2long_variation1.phpt
+++ b/ext/standard/tests/network/ip2long_variation1.phpt
@@ -141,19 +141,19 @@ bool(false)
 bool(false)
 
 --empty array--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --int indexed array--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --associative array--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --nested arrays--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --uppercase NULL--
@@ -184,7 +184,7 @@ bool(false)
 bool(false)
 
 --instance of classWithoutToString--
-Error: 2 - ip2long() expects parameter 1 to be string, object given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, object given, %s(%d)
 NULL
 
 --undefined var--
@@ -194,6 +194,6 @@ bool(false)
 bool(false)
 
 --resource--
-Error: 2 - ip2long() expects parameter 1 to be string, resource given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, resource given, %s(%d)
 NULL
 ===DONE===
diff --git a/ext/standard/tests/network/ip_x86_64.phpt b/ext/standard/tests/network/ip_x86_64.phpt
index 7b93815f92655dc3673c753ae09e636fc2f2aa38..6896063d6373dfca59362b424adfee84bacbe6b3 100644
--- a/ext/standard/tests/network/ip_x86_64.phpt
+++ b/ext/standard/tests/network/ip_x86_64.phpt
@@ -54,7 +54,7 @@ bool(false)
 bool(false)
 int(1869573999)
 
-Warning: ip2long() expects parameter 1 to be string, array given in %sip_x86_64.php on line %d
+Warning: ip2long() expects parameter 1 to be a valid path, array given in %sip_x86_64.php on line %d
 NULL
 
 Warning: long2ip() expects exactly 1 parameter, 0 given in %sip_x86_64.php on line %d
